### ISSUE TITLE
Question regarding directional derivative computation

### ISSUE THREAD

**Issue by hugoych** [tensorflow/tcav#41]
Hi,
While working with tcav code after reading the related paper " Interpretability Beyond Feature Attribution: Quantitative Testing with Concept Activation Vectors (TCAV)" (https://arxiv.org/pdf/1711.11279.pdf), I had trouble understanding the computation of the directional derivative.

From my understanding of the paper, 'Conceptual sensitivity' is measured by the directional derivative of  **logit** values with respect to activations at the layer of interest. Thus for scoring you take into account the proportions of image with positive directional derivative.

On the other hand in the code, as shown in the snippet below, sensitivity score seems to be  the directional derivative of **loss** values with respect to activations at the layer of interest.
https://github.com/tensorflow/tcav/blob/c7f55108e26c66fef9d21754d4d697ed6337ddc3/tcav/model.py#L54-L55

As the loss is a function of the logits, the gradient we obtain in the code depends of the loss function we choose. In your case with cross entropy  (snippet below) using  the article's notation we have :
![CodeCogsEqn-1](https://user-images.githubusercontent.com/29702745/61829566-33fa8080-ae37-11e9-9e73-6e5d71327b55.gif)

This leads to a change in magnitude ( which is not used is the score) but also in terms of sign.
https://github.com/tensorflow/tcav/blob/c7f55108e26c66fef9d21754d4d697ed6337ddc3/tcav/model.py#L173-L178

Though  it seems that in the code you compensate the sign change by counting the number of negative gradients instead (last snippet).
https://github.com/tensorflow/tcav/blob/c7f55108e26c66fef9d21754d4d697ed6337ddc3/tcav/tcav.py#L42-L60

Have I misunderstood either the paper or the code ? Or is there a specific reason to use loss instead of logit, as you still obtain the same score by changing sign ?
Thank you for your answer !

**Comment by hugoych** [comment]
Hi again,
@jameswex would you have any updates/explanations regarding the matter ? 

Thanks in advance

**Comment by jameswex** [comment]
@BeenKim can you please take a look? Thanks.

**Comment by MaximilianIdahl** [comment]
Hi,
I also stumbled across this issue. The pytorch re-implementation by @mlomnitz does indeed contain a commit in which this was adressed: https://github.com/mlomnitz/tcav_pytorch/commit/c2b5b968cc0c8c4fab1390328c7771f98c579224

Maybe @mlomnitz remembers what his thought process was when changing it? I know it has been a while though...

Thanks in advance

**Comment by hugoych** [comment]
@jameswex  Any update regarding the matter ? Thanks !

**Comment by BeenKim** [comment]
Sorry for this delayed response! We coded such that the two combination (loss and minus sign) would result in the same outcome. The code probably can be cleaned up to make this point clear.. sorry about that. Please reopen if your question remains!
